pysd/py_backend/output.py

```diff
@@ -0,0 +1,519 @@
+                callable(subclass.update) and
+                hasattr(subclass, 'postprocess') and
+                callable(subclass.postprocess) and
+                hasattr(subclass, 'add_run_elements') and
+                callable(subclass.add_run_elements) or
+                NotImplemented)
+
+    @abc.abstractmethod
+    def initialize(self, model, capture_elements):
+        """
+        Create the results object and its elements based on capture_elemetns.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update(self, model, capture_elements):
+        """
+        Update the results object at each iteration at which resutls are
+        stored.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def postprocess(self, **kwargs):
+        """
+        Perform different tasks at the time of returning the results object.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def add_run_elements(self, model, capture_elements):
+        """
+        Add elements with run cache to the results object.
+        """
+        raise NotImplementedError
+
+
+class DatasetHandler(OutputHandlerInterface):
+    """
+    Manages simulation results stored as netCDF4 Dataset.
+    """
+
+    def __init__(self, out_file):
+        self.out_file = out_file
+        self.step = 0
+        self.ds = None
+
+    def initialize(self, model, capture_elements):
+        """
+        Creates a netCDF4 Dataset and adds model dimensions and variables
+        present in the capture elements to it.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+
+        """
+        self.ds = nc.Dataset(self.out_file, "w")
+
+        # defining global attributes
+        self.ds.description = "Results for simulation run on" \
+            f"{t.ctime(t.time())} using PySD version {__version__}"
+        self.ds.model_file = model.py_model_file
+        self.ds.timestep = f"{model.components.time_step()}"
+        self.ds.initial_time = f"{model.components.initial_time()}"
+        self.ds.final_time = f"{model.components.final_time()}"
+
+        # creating variables for all model dimensions
+        for dim_name, coords in model.subscripts.items():
+            coords = np.array(coords)
+            # create dimension
+            self.ds.createDimension(dim_name, len(coords))
+            # length of the longest string in the coords
+            max_str_len = len(max(coords, key=len))
+
+            # create variable
+            # TODO: check if the type could be defined otherwise)
+            var = self.ds.createVariable(
+                dim_name, f"S{max_str_len}", (dim_name,))
+            # assigning values to variable
+            var[:] = coords
```

Comment:
It is preferable to define the coords as Dimensions and not Variables. The same you have done with time but using fixed values

Reply:
Is there a clear advantage in doing it as you suggest?

I just followed the netCDF4 documentation:

> The dimensions themselves are usually also defined as variables, called coordinate variables. The [Dataset.createVariable](https://unidata.github.io/netcdf4-python/#Dataset.createVariable) method returns an instance of the [Variable](https://unidata.github.io/netcdf4-python/#Variable) class whose methods can be used later to access and set variable data and attributes.

Reply:
I think that when opening the nc file with a netCDF viewer, e.g. ncview if you save the dimensions as Dimensions, you will be able to get each subscript value when iterating over a regular variable. Otherwise, I think you will have numerical indexes and you will need to go to the dimension variable to see the subscript name of that index. Saving dimensions as variables are used for non-regular spatial grids, but our case is not similar. Let me make some checks this evening and I will check if what I am saying is true.

Reply:
I have been checking the implementation of Dimension values and it is correct. ncview is not able to represent the values, I suppose that is because they are strings.

Reply:
Anyway, I'm implementing xr.Datasets now, so this will go to trash.